scheduler/reconnecting_picker/reconnecting_picker.go

```diff
@@ -0,0 +1,108 @@
+	if replacementIsNewer {
+		rp.logger.Debug("replacement has a newer version, keeping replacement")
+		return replacement
+	}
+
+	var picker func(*structs.Allocation, *structs.Allocation) *structs.Allocation
+
+	rs := ds.ReconcileStrategy()
+	rp.logger.Debug("picking according to strategy", "strategy", rs)
+
+	switch rs {
+	case structs.ReconcileOptionBestScore:
+		picker = rp.pickBestScore
+
+	case structs.ReconcileOptionKeepOriginal:
+		picker = rp.pickOriginal
+
+	case structs.ReconcileOptionKeepReplacement:
+		picker = rp.pickReplacement
+
+	case structs.ReconcileOptionLongestRunning:
+		picker = rp.pickLongestRunning
+	}
+
+	return picker(original, replacement)
+}
+
+// pickReconnectingAlloc returns the allocation to keep between the original
+// one that is reconnecting and one of its replacements.
+//
+// This function is not commutative, meaning that pickReconnectingAlloc(A, B)
+// is not the same as pickReconnectingAlloc(B, A). Preference is given to keep
+// the original allocation when possible.
+func (rp *ReconnectingPicker) pickBestScore(original *structs.Allocation, replacement *structs.Allocation) *structs.Allocation {
+
+	// Check if the replacement has better placement score.
+	// If any of the scores is not available, only pick the replacement if
+	// itself does have scores.
+	originalMaxScoreMeta := original.Metrics.MaxNormScore()
+	replacementMaxScoreMeta := replacement.Metrics.MaxNormScore()
+
+	replacementHasBetterScore := originalMaxScoreMeta == nil && replacementMaxScoreMeta != nil ||
+		(originalMaxScoreMeta != nil && replacementMaxScoreMeta != nil &&
+			replacementMaxScoreMeta.NormScore > originalMaxScoreMeta.NormScore)
+
+	// Check if the replacement has better client status.
+	// Even with a better placement score make sure we don't replace a running
+	// allocation with one that is not.
+	replacementIsRunning := replacement.ClientStatus == structs.AllocClientStatusRunning
+	originalNotRunning := original.ClientStatus != structs.AllocClientStatusRunning
+
+	if replacementHasBetterScore && (replacementIsRunning || originalNotRunning) {
+		return replacement
+	}
+
+	return original
+}
+
+func (rp *ReconnectingPicker) pickOriginal(original *structs.Allocation, _ *structs.Allocation) *structs.Allocation {
+	return original
+}
+
+func (rp *ReconnectingPicker) pickReplacement(_ *structs.Allocation, replacement *structs.Allocation) *structs.Allocation {
+	return replacement
+}
+
+func (rp *ReconnectingPicker) pickLongestRunning(original *structs.Allocation, replacement *structs.Allocation) *structs.Allocation {
+	lt := original.GetLeaderTask()
+
+	// Default to the first task in the group if no leader is found.
+	if lt.Name == "" {
+		lt = *original.Job.LookupTaskGroup(original.TaskGroup).Tasks[0]
+	}
+
+	// If the replacement has a later start time, keep the original.
+	if original.LatestStartOfTask(lt.Name).Sub(replacement.LatestStartOfTask(lt.Name)) < 0 {
+		return original
+	}
+
+	return replacement
```

Comment:
There are a few subtleties here that worry me a bit.

1. Task states are ephemeral, I think we only keep like 10 around? So for a long-living task I would expect the start event to have been long gone, specially if it uses features like `template` where each change will generate a handful of events. So in a more realistic scenario we would end up with `original.LatestStartOfTask` and `replacement.LatestStartOfTask` being both zero 😬 
2. This is only checking the leader task, or the first one, so it's not taking into consideration aspects such as task [lifecycle](https://developer.hashicorp.com/nomad/docs/job-specification/lifecycle). If the first task is a `prestart` sidecar, for example, I think this would give unexpected results.

I think a better approach would be use one of `LastRestart` or `StartedAt` (or probably a combination of both) from the alloc's `TaskState`?
https://github.com/hashicorp/nomad/blob/3193ac204f6564711004e00948e43146ce1399c4/nomad/structs/structs.go#L8816-L8826

We would also need to compare all tasks (or maybe at lest filter by `lifecycle`) so that we get a more general sense of what to consider "longest running" allocation.

Reply:
Ok, let me add a new filter to use a task from the main cycle :)

Reply:
Checking for main is good, but I think the biggest issue is relying on a specific task event. We can't really do that since we only store a limited number of them.

I think we need to rethink how we define longest running. The `TaskState` I mention could be helpful.

Reply:
Reimplemented! Now it uses the `StartAt`, `Restarts` and `LastRestart` fields.